File: src/test/mips.py
from asm.testsource import *  # pylint: disable=W0614

class MipsTestSource(TestSource):

    @property
    def name(self) -> str:
        return 'mips'

    @property
    def test_cases(self) -> TestCases:
        yield TestCase('should assemble single addi instruction', [
            self.make_call('addi', 'Reg::T1', 'Reg::T2', '0\'uint8')
        ], bytearray(b'\x00\x00\x49\x21'))

File: languages/cpp/test/mips.cpp
#ifndef ASM_ALL_TESTS
  #define CATCH_CONFIG_MAIN
#endif

#include <sstream>
#include "catch.hpp"
#include "../src/mips.cpp"

using Catch::Matchers::Equals;

using namespace std::string_literals;

TEST_CASE("mips tests", "[mips]") {
    std::ostringstream buf;

    SECTION("should assemble single addi instruction") {
        mips::addi(buf, mips::Reg::T1, mips::Reg::T2, 0);

        REQUIRE( buf.tellp() == 4 );
        REQUIRE_THAT(buf.str(), Equals("\x00\x00\x49\x21"s));
    }
}
